### PR TITLE
Added driver for HiFiBerry DAC+ ADC

### PR DESCRIPTION
**Pull request by hifiberry**
This adds a driver for the HiFiBerry DAC+ ADC which is a HAT that provides analog inputs and outputs.

### CODE REVIEW

**Comment by pelwell** [comment]
Leaving the whole issue of whether this can added to (or form the basis of) one of @timg236's megadrivers, you're going to have to justify enabling the sign-extension feature unilaterally (which will be an uphill battle). I suspect this would be better as an optional mode, and pushed upstream.

**Comment by hifiberry** [comment]
About the megadrivers: Right now the DAC+ driver (which is the base for this) is not included in the new megadrive, therefore I think it's good to have it still separately at the moment.

**Comment by j-schambacher** [comment]
My comment about the sign extension flag: as long as the I2S interface runs with a bitclk of 64x everything works correctly. But with a 48x bitclk (some ADCs offer that) the I2S input shift register fills the PCM buffer with 8bit random data + 24bit valid data into the 32bit words. The sign extension will just correct the 8bit random data by copying the 24th bit into bits 25-32. In all other modes 32 or 16 bits it has no effect. Output is also not effected. 
Nevertheless, as this DAC+ ADC driver/sound card runs at 64x we can leave it out for now.

**Comment by timg236** [comment]
I have a work in progress composite PCM512X driver which currently supports
allo-piano-dac, dionaudio-loco-v2, justboom-dac, iqaudio-dac

I'll probably try to merge that first as soon as I get chance to test it. The HifiBerry DAC+ ADC is more complex so it might be better to keep it as a separate driver anyway.

**Comment by pelwell** [comment]
Without the sign-extension patch this becomes much more straight-forward. I've patched up the README, and if you can address the coding-style issues reported by checkpatch (don't worry about the requests for documentation, but a Signed-off-by: in the commit would be nice) we can get this merged - just an additional commit with the fixes and I'll squash it at merge time.

File: sound/soc/codecs/pcm1861.c [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	int ret;
+    	ret = devm_snd_soc_register_component(&pdev->dev,
+					&soc_component_dev_pcm1861,
+					&pcm1861_dai, 1);
+	return ret;
+}
+
+static int pcm1861_remove(struct platform_device *pdev)
+{
+	snd_soc_unregister_component(&pdev->dev);
```
Comment (HiassofT) [inline sound/soc/codecs/pcm1861.c]: this should be dropped as the component is registered via devm

**Comment by pelwell** [comment]
Any progress on this? You'll need to rebase against the latest commits to the branch.

**Comment by hifiberry** [comment]
> Any progress on this? You'll need to rebase against the latest commits to the branch.

I should be able to find the time for this in the next days.

**Comment by hifiberry** [comment]
As always I'm having trouble with rebase/merging. @pelwell Can you help with this?

**Comment by pelwell** [comment]
Giving definitive instructions is hard without knowing how you've configured your remotes, but assuming that `origin` is `raspberrypi/linux` and `hifiberry` is `hifiberry/linux`:
```bash
$ git fetch origin
$ git rebase origin/rpi-4.18.y
$ git push -f hifiberry HEAD:rpi-4.18.y
```

**Comment by pelwell** [comment]
I feel you shouldn't need the pcm1861 driver, which is essentially just a way of presenting this information:
```
	.name = "pcm1861-hifi",
	.capture = {
		.channels_min = 2,
		.channels_max = 2,
		.rates = SNDRV_PCM_RATE_CONTINUOUS,
		.formats = SNDRV_PCM_FMTBIT_S16_LE |
			   SNDRV_PCM_FMTBIT_S24_LE |
			   SNDRV_PCM_FMTBIT_S32_LE,
	},
```
The dmic codec driver, if you ignore the enable GPIO and wakeup delay, presents the same information to ALSA, but with different stream names.

**Comment by hifiberry** [comment]
> I feel you shouldn't need the pcm1861 driver, which is essentially just a way of presenting this information:

Dmic has channels_min = 1 and channels_max = 8 which wouldn't work here.

**Comment by pelwell** [comment]
It's configurable by DT properties. From [the bindings](https://github.com/raspberrypi/linux/blob/rpi-4.18.y/Documentation/devicetree/bindings/sound/dmic.txt):
```
Optional properties:
	- dmicen-gpios: GPIO specifier for dmic to control start and stop
	- num-channels: Number of microphones on this DAI
	- wakeup-delay-ms: Delay (in ms) after enabling the DMIC

Example node:

	dmic_codec: dmic@0 {
		compatible = "dmic-codec";
		dmicen-gpios = <&gpio4 3 GPIO_ACTIVE_HIGH>;
		num-channels = <1>;
		wakeup-delay-ms <50>;
	};
```

**Comment by hifiberry** [comment]
Ok, I see. If this works better for you, we'll look into this. This will need some changes in the driver and the device tree overlay. We'l look into this.

**Comment by HiassofT** [comment]
channels min 1/max 8 isn't a problem, ASoC intersects constraints of all involved parties (codec, platform, machine driver)

**Comment by pelwell** [comment]
It's one fewer file to maintain, which has to be a good thing. I'm happy to help with the required changes, seeing as this is a late request.

**Comment by hifiberry** [comment]
> It's one fewer file to maintain, which has to be a good thing. I'm happy to help with the required changes, seeing as this is a late request.

Ok, as I also get a lot of conflicts during rebase, I think it's the easiest way to create a new pull request based on the dmic driver.

**Comment by pelwell** [comment]
You can skip the rebase conflicts by temporarily creating a new branch and replaying your changes on top of it. Assuming you have `n` downstream commits in your rpi-4.18.y you want to rebase onto the new upstream branch:
```
# Starting on your old rpi-4.18.y
$ git checkout -b temp origin/rpi-4.18.y
$ git rebase --onto temp rpi-4.18.y~7 rpi-4.18.y
$ git reset --hard temp
$ git branch -D temp
```
Which can also be written as:
```
# Starting on your old rpi-4.18.y
$ git branch temp
$ git reset --hard origin/rpi-4.18.y
$ git rebase --onto rpi-4.18.y temp~7 temp
$ git branch -D temp
```
The temporary branch is just a convenient alias for a commit - the top of either the new or old tree.

**Comment by hifiberry** [comment]
@pelwell: Can't get the rebase working, always getting conflicts. However with the new dmic driver, there have been quite some changes again. Therefore I think, it's the easiest way to close this and open a new pull request. I'll do this now. Thanks.

**Comment by pelwell** [comment]
You can do that if you want, but it isn't necessary. Prepare your tree however you want - start again from scratch, fix the rebasing. etc. - it doesn't matter. When you are ready to go with the new PR, just use the force push option to update your branch that forms the basis of this PR:
```
git push -f hifiberry HEAD:rpi-4.18.y
```
This will automatically update the rpi PR.

**Comment by hifiberry** [comment]
Just pushed the updated PR based with the ADC part based on the dmic driver.

**Comment by pelwell** [comment]
That looks good to me (after rebasing and patching up the README again). If you reply with a `Signed-off-by:` line I can add it to the commit message during the merge.

**Comment by hifiberry** [comment]
Signed-off-by: Daniel Matuschek <daniel@hifiberry.com>

**Comment by hifiberry** [comment]
Thanks for the support!

**Comment by plbossart** [comment]
FWIW, the use of the dmic-codec for this driver seems totally overkill. In cases where the codec is not configurable via SPI/I2C, or does not need to be configured, using soc-dummy-codec and soc-dummy-dai are the recommended solutions.

**Comment by hifiberry** [comment]
@pelwell Just noticed that LibreElec already uses 4.19 and the patch is only in 4.18. Can you integrate the patch also into 4.19?

**Comment by pelwell** [comment]
Done. Updates for the newer kernels are in progress.

**Pull request merged**